database/pgsql/migrations/00001_initial_schema.go

```diff
@@ -14,179 +14,220 @@
-func init() {
-	RegisterMigration(migrate.Migration{
-		ID: 1,
-		Up: migrate.Queries([]string{
+var (
+	// entities are the basic building blocks to relate the vulnerabilities with
+	// the ancestry.
+	entities = dbutil.MigrationQuery{
+		Up: []string{
 			// namespaces
 			`CREATE TABLE IF NOT EXISTS namespace (
-		id SERIAL PRIMARY KEY,
-		name TEXT NULL,
-		version_format TEXT,
-		UNIQUE (name, version_format));`,
+				id SERIAL PRIMARY KEY,
+				name TEXT NULL,
+				version_format TEXT,
+				UNIQUE (name, version_format));`,
 			`CREATE INDEX ON namespace(name);`,
 
 			// features
 			`CREATE TABLE IF NOT EXISTS feature (
-		id SERIAL PRIMARY KEY,
-		name TEXT NOT NULL,
-		version TEXT NOT NULL,
-		version_format TEXT NOT NULL,
-		UNIQUE (name, version, version_format));`,
+				id SERIAL PRIMARY KEY,
+				name TEXT NOT NULL,
+				version TEXT NOT NULL,
+				version_format TEXT NOT NULL,
+				UNIQUE (name, version, version_format));`,
 			`CREATE INDEX ON feature(name);`,
 
 			`CREATE TABLE IF NOT EXISTS namespaced_feature (
-		id SERIAL PRIMARY KEY,
-		namespace_id INT REFERENCES namespace,
-		feature_id INT REFERENCES feature,
-		UNIQUE (namespace_id, feature_id));`,
-
+				id SERIAL PRIMARY KEY,
+				namespace_id INT REFERENCES namespace,
+				feature_id INT REFERENCES feature,
+				UNIQUE (namespace_id, feature_id));`,
+		},
+		Down: []string{
+			`DROP TABLE IF EXISTS namespace, feature, namespaced_feature CASCADE;`,
+		},
+	}
+
+	// detector is analysis extensions used by the worker.
+	detector = dbutil.MigrationQuery{
+		Up: []string{
+			// Detector Type
+			`CREATE TYPE detector_type AS ENUM ('namespace', 'feature')`,
+
+			// Detector
+			`CREATE TABLE IF NOT EXISTS detector (
+				id SERIAL PRIMARY KEY,
+				name TEXT NOT NULL,
+				version TEXT NOT NULL,
+				type detector_type NOT NULL,
+				UNIQUE (name, version, type));`,
+		},
+		Down: []string{
+			`DROP TABLE IF EXISTS detector CASCADE;`,
+			`DROP TYPE IF EXISTS detector_type;`,
+		},
+	}
+
+	// layer contains all metadata and scanned features and namespaces.
+	layer = dbutil.MigrationQuery{
+		Up: []string{
 			// layers
 			`CREATE TABLE IF NOT EXISTS layer(
-		id SERIAL PRIMARY KEY, 
-		hash TEXT NOT NULL UNIQUE);`,
+				id SERIAL PRIMARY KEY,
+				hash TEXT NOT NULL UNIQUE);`,
+
+			`CREATE TABLE IF NOT EXISTS layer_detector(
+				id SERIAL PRIMARY KEY,
+				layer_id INT REFERENCES layer ON DELETE CASCADE,
+				detector_id INT REFERENCES detector ON DELETE CASCADE,
+				UNIQUE(layer_id, detector_id));`,
+			`CREATE INDEX ON layer_detector(layer_id);`,
 
 			`CREATE TABLE IF NOT EXISTS layer_feature (
-		id SERIAL PRIMARY KEY, 
-		layer_id INT REFERENCES layer ON DELETE CASCADE, 
-		feature_id INT REFERENCES feature ON DELETE CASCADE,
-		UNIQUE (layer_id, feature_id));`,
+				id SERIAL PRIMARY KEY,
+				layer_id INT REFERENCES layer ON DELETE CASCADE, 
+				feature_id INT REFERENCES feature ON DELETE CASCADE,
+				detector_id INT REFERENCES detector ON DELETE CASCADE,
```

Comment:
Layer feature now has reference to detector, which is used to produce this feature. 

The reason for having detector here instead of in `Feature` table is that:
1. Detector only make sense in the context of layer: `Feature`, `Namespace`, and `NamespacedFeature` are also used by the vulnerabilities, which doesn't really care about what scans a feature.
2. Suppose: 
Detector1, Detector2 scans Layer1 and found feature Feature1 by Detector1
Detector1, Detector2 scans Layer2 and found feature Feature1 by Detector2
Since Feature1 is found by Detector1 in Layer1, this relation should be stored somewhere in the database. I think this is the best place.